### ISSUE TITLE
alerting.engine: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by safa-topal** [grafana/grafana#47339]
**What happened**:
After upgrading Grafana to 8.3.2 from 8.2.5, process is going into panic mode and goes down.

Many logs like these before going into panic mode(may not be really related):
`
lvl=info msg="Alert Rule returned no data" logger=alerting.evalContext ruleId=2473 name="1129 INSUFFICIENT_MPPT_COUNT" changing state to=keep_state
lvl=info msg="Alert Rule returned no data" logger=alerting.evalContext ruleId=2467 name="1129 SYSTEM_TEMPERATURE_LOW" changing state to=keep_state
lvl=info msg="Alert Rule returned no data" logger=alerting.evalContext ruleId=2580 name="1137 CHARGE_CURRENT_HIGH" changing state to=keep_state
lvl=info msg="Alert Rule returned no data" logger=alerting.evalContext ruleId=2723 name="1126 RECTIFIER_COMM_ERROR" changing state to=keep_state
`

**Here's the final log. It seems about going into panic mode and then service becomes unresponsive:**
`
lvl=eror msg="Alert Panic" logger=alerting.engine error="runtime error: invalid memory address or nil pointer dereference" stack="[engine.go:189 panic.go:1038 panic.go:221 signal_unix.go:735 <autogenerated>:1 sql.go:1268 sql.go:1695 sql.go:1674 db.go:121 session_raw.go:74 session_get.go:103 session_get.go:90 session_get.go:23 datasource.go:28 session.go:67 session.go:59 datasource.go:22 service.go:77 value.go:543 value.go:339 bus.go:119 bus.go:231 query.go:143 query.go:53 eval_handler.go:37 engine.go:200]"
`

**What you expected to happen**:
Grafana working as expected and it doesn't go down

**How to reproduce it (as minimally and precisely as possible)**:
Not too sure, either amount of alerts or particular types of alerts not working well with the alerting.engine

**Anything else we need to know?**:
It was working well before the upgrade and only major change in Grafana seems alerting related changes.
Similar issues:
* https://github.com/grafana/grafana/issues/47093
* https://github.com/grafana/grafana/issues/47155

**Environment**:
- Grafana version: 8.3.2
- Data source type & version: - 
- OS Grafana is installed on: AWS
- User OS & Browser: Fedora
- Grafana plugins: -
- Others: -

**Comment by mellieA** [comment]
@safa-topal are you able to tell from logs or timing if there are any specific alerts firing or evaluating at the time that this happens?

**Comment by safa-topal** [comment]
@mellieA yes, indeed. Before the failure I can see that many alerting related evaluations are being executed.

**Comment by safa-topal** [comment]
@mellieA FYI, user of this Grafana instance is using Legacy alerting

**Comment by armandgrillet** [comment]
https://github.com/grafana/grafana/pull/46997 should fix the issue in 8.4, 8.5, and forward.

**Comment by MichaHoffmann** [comment]
> #46997 should fix the issue in 8.4, 8.5, and forward.

Not sure about that tbh. The call trace was ending in database/sql here not in gonum.

**Comment by mhoffm-aiven** [comment]
testing to call 
```
sqlStore.GetDataSource(nil, &query)
```
the stack trace matches from datasource.go:22 to sql.go:1268:
```
	/home/michael.hoffmann/sdk/go1.17.3/src/database/sql/sql.go:1268 +0xdc
database/sql.(*DB).query(0x0, {0x0, 0x0}, {0xc000517180, 0x13d}, {0xc000a69860, 0x2, 0x2}, 0x20)
	/home/michael.hoffmann/sdk/go1.17.3/src/database/sql/sql.go:1695 +0x5d
database/sql.(*DB).QueryContext(0x0, {0x0, 0x0}, {0xc000517180, 0x13d}, {0xc000a69860, 0x2, 0x2})
	/home/michael.hoffmann/sdk/go1.17.3/src/database/sql/sql.go:1674 +0xdf
xorm.io/core.(*DB).QueryContext(0xc00004d400, {0x0, 0x0}, {0xc000517180, 0x2}, {0xc000a69860, 0x0, 0x15cdd00})
	/home/michael.hoffmann/go/pkg/mod/xorm.io/core@v0.7.3/db.go:121 +0x45
xorm.io/xorm.(*Session).queryRows(0xc000300a80, {0xc000517180, 0x13d}, {0xc000a69860, 0xc000387640, 0x111ae80})
	/home/michael.hoffmann/go/pkg/mod/xorm.io/xorm@v0.8.2/session_raw.go:74 +0x6ca
xorm.io/xorm.(*Session).nocacheGet(0xbb2246, 0x19, 0x0, {0x10ea1e0, 0xc0000d9c20}, {0xc000517180, 0x13d}, {0xc000a69860, 0x2, 0x2})
	/home/michael.hoffmann/go/pkg/mod/xorm.io/xorm@v0.8.2/session_get.go:103 +0xe5
xorm.io/xorm.(*Session).get(0xc000300a80, {0x10ea1e0, 0xc0000d9c20})
	/home/michael.hoffmann/go/pkg/mod/xorm.io/xorm@v0.8.2/session_get.go:90 +0x6e7
xorm.io/xorm.(*Session).Get(0x7f7b0193b108, {0x10ea1e0, 0xc0000d9c20})
	/home/michael.hoffmann/go/pkg/mod/xorm.io/xorm@v0.8.2/session_get.go:23 +0x78
github.com/grafana/grafana/pkg/services/sqlstore.(*SQLStore).GetDataSource.func1(0xc000a69780)
	/home/michael.hoffmann/git/grafana/pkg/services/sqlstore/datasource.go:28 +0xed
github.com/grafana/grafana/pkg/services/sqlstore.withDbSession({0x0, 0x0}, 0xc00052e000, 0xc000189da0)
	/home/michael.hoffmann/git/grafana/pkg/services/sqlstore/session.go:67 +0x128
github.com/grafana/grafana/pkg/services/sqlstore.(*SQLStore).WithDbSession(...)
	/home/michael.hoffmann/git/grafana/pkg/services/sqlstore/session.go:59
github.com/grafana/grafana/pkg/services/sqlstore.(*SQLStore).GetDataSource(0xc000168300, {0x0, 0x0}, 0xc000189e38)
	/home/michael.hoffmann/git/grafana/pkg/sqlStore.GetDataSource(nil, &query)                                                     services/sqlstore/datasource.go:22 +0x7b
```

so im pretty sure that somehow on version 8.3.2 a nil context sneaks into the call to `GetDataSource`

**Comment by safa-topal** [comment]
Upgrading to 8.4.6 didn't fix the issue.

**Comment by EusebiusAditya** [comment]
is there any update regarding this issue, find a similar issue when using 8.3.2

**Comment by yuri-tceretian** [comment]
I do not think this is related to alerting. Yes, it might be a symptom of the problem because alerting is the most active service inside Grafana. 
The reason I think this is that alerting code has recovery from panic:

https://github.com/grafana/grafana/blob/5412a903b1df60fe739316e827cabbc758c3325c/pkg/services/alerting/engine.go#L187-L198

https://github.com/grafana/grafana/blob/5412a903b1df60fe739316e827cabbc758c3325c/pkg/services/alerting/engine.go#L172-L176

and the mentioned log message was emitted by one of those logs (I think it was the former).

At last, to prove that I simulated failure by updating the code and providing nil context down to evaluation:

https://github.com/grafana/grafana/blob/5412a903b1df60fe739316e827cabbc758c3325c/pkg/services/alerting/engine.go#L184

I could see pretty much the same errors as mentioned above but the process did not crash.

**Comment by bogdanioanliviu** [comment]
Same error happening in 9.2.7 and 9.3.0 but not in 9.2.5
`level=error msg="Request error" error="runtime error: invalid memory address or nil pointer dereference" stack="/usr/local/go/src/runtime/panic.go:260 (0x451315)\n/usr/local/go/src/runtime/signal_unix.go:835 (0x4512e5)\n/drone/src/pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:202 (0x24bb735)\n/drone/src/pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:72 (0x24b9b26)\n/drone/src/pkg/tsdb/azuremonitor/azuremonitor.go:163 (0x24c70c5)\n/opt/drone/gomodcache/github.com/grafana/grafana-plugin-sdk-go@v0.139.0/backend/data.go:33 (0xb733ae)\n/opt/drone/gomodcache/github.com/grafana/grafana-plugin-sdk-go@v0.139.0/backend/datasource/query_type_mux.go:85 (0x132eee1)\n/drone/src/pkg/tsdb/azuremonitor/azuremonitor.go:50 (0x24c6066)\n/drone/src/pkg/plugins/backendplugin/coreplugin/core_plugin.go:81 (0x2ac8a90)\n/drone/src/pkg/plugins/plugins.go:256 (0x1144254)\n/drone/src/pkg/plugins/manager/client/client.go:36 (0x2ae7db1)\n/drone/src/pkg/plugins/backendplugin/instrumentation/instrumentation.go:32 (0x2ae70f7)\n/drone/src/pkg/plugins/backendplugin/instrumentation/instrumentation.go:61 (0x2ae78a4)\n/drone/src/pkg/plugins/manager/client/client.go:35 (0x2ae76d9)\n/drone/src/pkg/services/query/query.go:211 (0x1ec65e2)\n/drone/src/pkg/services/query/query.go:79 (0x1ec513c)\n/drone/src/pkg/api/metrics.go:57 (0x2439914)\n/drone/src/pkg/api/routing/routing.go:17 (0x167fe21)\n/drone/src/pkg/api/response/web_hack.go:40 (0x1171115)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:135 (0xff42b4)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/csp.go:22 (0x1677032)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/services/contexthandler/contexthandler.go:175 (0x1662401)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/render.go:44 (0xff4e75)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/macaron.go:141 (0xff435b)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/csrf/csrf.go:64 (0x2177445)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/recovery.go:171 (0x167a7c1)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/logger.go:40 (0x16781bc)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/request_metrics.go:58 (0x167b5c9)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/middleware/request_tracing.go:88 (0x167cd1a)\n/usr/local/go/src/net/http/server.go:2109 (0x712b2e)\n/drone/src/pkg/web/context.go:44 (0xff2dc2)\n/drone/src/pkg/web/router.go:155 (0xff63ed)\n/drone/src/pkg/web/router.go:211 (0xff7055)\n/drone/src/pkg/web/macaron.go:167 (0xff47d0)\n/usr/local/go/src/net/http/server.go:2947 (0x715f2b)\n/usr/local/go/src/net/http/server.go:1991 (0x711606)\n/usr/local/go/src/runtime/asm_amd64.s:1594 (0x46f080)\n"`

**Comment by safa-topal** [comment]
The user's issue was resolved at the end, although didn't get feedback from this user how their issue was resolved. 

> Also, does this bug happen in Grafana 9 and later?

No, at least didn't hear from any user that it happens

**Comment by andresmgot** [comment]
> Same error happening in 9.2.7 and 9.3.0 but not in 9.2.5 `level=error msg="Request error" error="runtime error: invalid memory address or nil pointer dereference" stack="/usr/local/go/src/runtime/panic.go:260 (0x451315)\n/usr/local/go/src/runtime/signal_unix.go:835 (0x4512e5)\n/drone/src/pkg/tsdb/azuremonitor/resourcegraph/azure-resource-graph-datasource.go:202 

Thanks for reporting that @bogdanioanliviu, it was caused by an empty response. Should be fixed by https://github.com/grafana/grafana/pull/59691

**Comment by armandgrillet** [comment]
The fix above has been backported to 9.2 and 9.3, next releases will thus include the fix.
